src/bucket.ts

```diff
@@ -3174,6 +3185,201 @@ class Bucket extends ServiceObject {
+  uploadDirectory(
+    directoryPathString: string,
+    options: UploadDirectoryOptions,
+    callback: UploadDirectoryCallback
+  ): void;
+  uploadDirectory(
+    directoryPathString: string,
+    callback: UploadDirectoryCallback
+  ): void;
+  /**
+   * @typedef {object} UploadDirectoryOptions Configuration options for Bucket#uploadDirectory().
+   * @param {boolean} [options.recurse] Whether to recursevely upload all files from
+   *     all subdirectories
+   * @param {boolean} [options.gzip] Automatically gzip every file. This will set
+   *     `options.metadata.contentEncoding` to `gzip`.
+   * @param {object} [options.metadata] See an
+   *     [Objects: insert request
+   * body](https://cloud.google.com/storage/docs/json_api/v1/objects/insert#request_properties_JSON).
+   * @param {string} [options.offset] The starting byte of the upload stream, for
+   *     resuming an interrupted upload. Defaults to 0.
+   * @param {string} [options.predefinedAcl] Apply a predefined set of access
+   *     controls to every object.
+   *
+   *     Acceptable values are:
+   *     - **`authenticatedRead`** - Object owner gets `OWNER` access, and
+   *       `allAuthenticatedUsers` get `READER` access.
+   *
+   *     - **`bucketOwnerFullControl`** - Object owner gets `OWNER` access, and
+   *       project team owners get `OWNER` access.
+   *
+   *     - **`bucketOwnerRead`** - Object owner gets `OWNER` access, and project
+   *       team owners get `READER` access.
+   *
+   *     - **`private`** - Object owner gets `OWNER` access.
+   *
+   *     - **`projectPrivate`** - Object owner gets `OWNER` access, and project
+   *       team members get access according to their roles.
+   *
+   *     - **`publicRead`** - Object owner gets `OWNER` access, and `allUsers`
+   * get `READER` access.
+   * @param {boolean} [options.private] Make the uploaded files private. (Alias for
+   *     `options.predefinedAcl = 'private'`)
+   * @param {boolean} [options.public] Make the uploaded files public. (Alias for
+   *     `options.predefinedAcl = 'publicRead'`)
+   * @param {boolean} [options.resumable] Force a resumable upload. (default:
+   *     true for files larger than 5 MB).
+   * @param {string} [options.uri] The URI for an already-created resumable
+   *     upload. See {@link File#createResumableUpload}.
+   * @param {string} [options.userProject] The ID of the project which will be
+   *     billed for the request.
+   * @param {string|boolean} [options.validation] Possible values: `"md5"`,
+   *     `"crc32c"`, or `false`. By default, data integrity is validated with an
+   *     MD5 checksum for maximum reliability. CRC32c will provide better
+   *     performance with less reliability. You may also choose to skip
+   * validation completely, however this is **not recommended**.
+   */
+  /**
+   * @typedef {array} UploadDirectoryResponse
+   * @property {object} [response] Response for each of the uploaded files.
```

Comment:
I think since we're describing an array these should be index values.

Reply:
done